src/Lean/Elab/Coinductive.lean

```diff
@@ -0,0 +1,467 @@
+      let rhs := mkConst info.name levels
+      let rhs := mkAppN rhs (params ++ calls ++ args)
+
+      let goalType ← mkEq lhs rhs
+      let goal ← mkFreshExprMVar goalType
+
+      let goalMVarId := goal.mvarId!
+
+      let .some #[fixEq] ←  getEqnsFor? (removeFunctorPostfix info.name) | throwError "did not generate unfolding theorem"
+      let existentialEquiv := mkConst (info.name ++ `existential_equiv) levels
+
+      let mut fixEq := mkConst fixEq levels
+      fixEq := mkAppN fixEq params
+      for arg in args do
+        fixEq ← mkCongrFun fixEq arg
+
+      let newGoal ← rewriteGoalUsingEq goalMVarId existentialEquiv
+      newGoal.assign fixEq
+
+      let goal ← instantiateMVars goal
+      mkLambdaFVars (params ++ args) goal
+    trace[Elab.coinductive] "res: {res}"
+    addDecl <| .defnDecl (←mkDefinitionValInferringUnsafe
+      ((removeFunctorPostfix info.name) ++ `functor_unfold) info.levelParams (←inferType res) res .opaque)
+
+/--
+  Generates a constructor for a coinductive predicate that corresponds to constructors
+  in the original `InductiveView`.
+
+  The process:
+  1. Takes the flat inductive constructor type
+  2. Fills recursive call parameters with the actual coinductive predicates
+  3. Converts to existential form using the equivalence lemma
+  4. Applies the unrolling rule to get the final constructor form
+-/
+private def generateCoinductiveConstructor (infos : Array InductiveVal) (ctorSyntax : Syntax)
+    (numParams : Nat) (name : Name) (ctor : ConstructorVal) : TermElabM Unit := do
+  trace[Elab.coinductive] "Generating constructor: {removeFunctorPostfixInCtor ctor.name}"
+  let numPreds := infos.size
+  let predNames := infos.map fun val => removeFunctorPostfix val.name
+  let levelParams := infos[0]!.levelParams.map mkLevelParam
+  /-
+    We start by looking at the type of the constructor of the flat inductive and then by introducing
+    all its parameters to the scope.
+  -/
+  forallBoundedTelescope ctor.type (numParams + numPreds) fun args body => do
+    /-
+      The first `numParams` many items of `args` are parameters from the original definition,
+      while the remaining ones are free variables that correspond to recursive calls.
+    -/
+    let params := args.take numParams
+    let predFVars := args.extract numParams
+    /-
+      We will fill recursive calls in the body with the just defined (co)inductive predicates.
+    -/
+    let mut predicates : Array Expr := predNames.map (mkConst · levelParams)
+    predicates := predicates.map (mkAppN · params)
+    let body := body.replaceFVars predFVars predicates
+    /-
+      Now, we look at the rest of the constructor.
+      We start by collecting its non-parameter premises, as well as inspecting its conclusion.
+    -/
+    let res ← forallTelescope body fun bodyArgs bodyExpr => do
+      /-
+        First, we look at conclusion and pick out all arguments that are non-parameters.
```

Comment:
“fields” is a good name for the arguments of a constructor that are not the parameters of the whole declaration.